modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C

```diff
@@ -0,0 +1,700 @@
+  const Real dlbl = (b >= 0.0 ? 1.0 : -1.0); // d(lbl)/db
+  switch (_flux_limiter_type)
+  {
+    case FluxLimiterTypeEnum::MinMod:
+    {
+      if (lal <= lbl)
+      {
+        limited = s * lal;
+        dlimited_db = 0.0;
+      }
+      else
+      {
+        limited = s * lbl;
+        dlimited_db = s * dlbl;
+      }
+      return;
+    }
+    case FluxLimiterTypeEnum::VanLeer:
+    {
+      limited = s * 2 * lal * lbl / (lal + lbl);
+      dlimited_db = s * 2 * lal * (dlbl / (lal + lbl) - lbl * dlbl / std::pow(lal + lbl, 2));
+      return;
+    }
+    case FluxLimiterTypeEnum::MC:
+    {
+      const Real av = 0.5 * std::abs(a + b);
+      if (2 * lal <= av && lal <= lbl)
+      {
+        // 2 * lal is the smallest
+        limited = s * 2.0 * lal;
+        dlimited_db = 0.0;
+      }
+      else if (2 * lbl <= av && lbl <= lal)
+      {
+        // 2 * lbl is the smallest
+        limited = s * 2.0 * lbl;
+        dlimited_db = s * 2.0 * dlbl;
+      }
+      else
+      {
+        // av is the smallest
+        limited = s * av;
+        // if (a>0 and b>0) then d(av)/db = 0.5 = 0.5 * dlbl
+        // if (a<0 and b<0) then d(av)/db = -0.5 = 0.5 * dlbl
+        // if a and b have different sign then limited=0, above
+        dlimited_db = s * 0.5 * dlbl;
+      }
+      return;
+    }
+    case FluxLimiterTypeEnum::superbee:
+    {
+      const Real term1 = std::min(2.0 * lal, lbl);
+      const Real term2 = std::min(lal, 2.0 * lbl);
+      if (term1 >= term2)
+      {
+        if (2.0 * lal <= lbl)
+        {
+          limited = s * 2 * lal;
+          dlimited_db = 0.0;
+        }
+        else
+        {
+          limited = s * lbl;
+          dlimited_db = s * dlbl;
+        }
+      }
+      else
+      {
+        if (lal <= 2.0 * lbl)
+        {
+          limited = s * lal;
+          dlimited_db = 0.0;
+        }
+        else
+        {
+          limited = s * 2.0 * lbl;
+          dlimited_db = s * 2.0 * dlbl;
+        }
+      }
+      return;
+    }
+    default:
+      return;
+  }
+}
+
+Real
+AdvectiveFluxCalculatorBase::getKij(dof_id_type node_i, dof_id_type node_j) const
+{
+
+  const auto & row_find = _kij.find(node_i);
+  mooseAssert(row_find != _kij.end(),
+              "AdvectiveFluxCalculatorBase UserObject " << name() << " Kij does not contain node "
+                                                        << node_i);
+  const std::map<dof_id_type, Real> & kij_row = row_find->second;
+  const auto & entry_find = kij_row.find(node_j);
+  mooseAssert(entry_find != kij_row.end(),
+              "AdvectiveFluxCalculatorBase UserObject " << name() << " Kij on row " << node_i
+                                                        << " does not contain node " << node_j);
+
+  return entry_find->second;
+}
+
+std::map<dof_id_type, Real>
+AdvectiveFluxCalculatorBase::getdFluxOutdu(dof_id_type node_i) const
+{
+  const auto & row_find = _dflux_out_du.find(node_i);
+  mooseAssert(row_find != _dflux_out_du.end(),
```

Comment:
I'm not sure about this, @andrsd.  I don't think the user can ever trigger this Assert by making a weirdo input file or mesh.  I think the only way of triggering the Assert is by poor C++ coding, ie, mainly the Assert is for me alone.  So i think an Assert is better than something that's going to be in the optimized code.   Do you agree?

Reply:
I was thinking that for the case when people (developers) will be using your API, it would be nice to guard it, because they might get into weird problems. So it is better to error out where the actual error is. We had a similar problem in RELAP and the code was bombing out in a different spot and you where not able to tell. So I like to error out as soon as possible.

However, if you do not think it is a problem, keep it this way.

Reply:
I think your experience trumps my belief